daemon/stringset.c

```diff
@@ -0,0 +1,337 @@
+
+#include "stringset.h"
+
+#include "logging.h"
+
+/* ========================================================================= *
+ * Prototypes
+ * ========================================================================= */
+
+/* ------------------------------------------------------------------------- *
+ * UTILITY
+ * ------------------------------------------------------------------------- */
+
+static void   stringset_ctor       (stringset_t *self);
+static void   stringset_dtor       (stringset_t *self);
+stringset_t  *stringset_create     (void);
+void          stringset_delete     (stringset_t *self);
+void          stringset_delete_at  (stringset_t **pself);
+void          stringset_delete_cb  (void *self);
+guint         stringset_size       (const stringset_t *self);
+bool          stringset_empty      (const stringset_t *self);
+const GList  *stringset_list       (const stringset_t *self);
+bool          stringset_has_item   (const stringset_t *self, const gchar *item);
+bool          stringset_add_item   (stringset_t *self, const gchar *item);
+bool          stringset_remove_item(stringset_t *self, const gchar *item);
+bool          stringset_clear      (stringset_t *self);
+GVariant     *stringset_to_variant (const stringset_t *self);
+gchar        *stringset_to_string  (const stringset_t *self);
+gchar       **stringset_to_strv    (const stringset_t *self);
+stringset_t  *stringset_from_strv  (char **vector);
+stringset_t  *stringset_copy       (const stringset_t *self);
+void          stringset_swap       (stringset_t *self, stringset_t *that);
+stringset_t  *stringset_filter_out (const stringset_t *self, const stringset_t *mask);
+stringset_t  *stringset_filter_in  (const stringset_t *self, const stringset_t *mask);
+bool          stringset_equal      (const stringset_t *self, const stringset_t *that);
+bool          stringset_extend     (stringset_t *self, const stringset_t *that);
+bool          stringset_assign     (stringset_t *self, const stringset_t *that);
+
+/* ========================================================================= *
+ * STRINGSET
+ * ========================================================================= */
+
+struct stringset_t
+{
+    /* GQueue keeps the set in deterministic order */
+    GQueue      sst_list;
+
+    /* GHashTable is used for lookups and memory management */
+    GHashTable *sst_hash;
+};
+
+static void
+stringset_ctor(stringset_t *self)
+{
+    //log_debug("stringset() create");
+    g_queue_init(&self->sst_list);
+    self->sst_hash = g_hash_table_new_full(g_str_hash, g_str_equal,
+                                           g_free, NULL);
+}
+
+static void
+stringset_dtor(stringset_t *self)
+{
+    //log_debug("stringset() delete");
+
+    stringset_clear(self);
+
+    if( self->sst_hash ) {
+        g_hash_table_unref(self->sst_hash),
+            self->sst_hash = NULL;
+    }
+}
+
+stringset_t *
+stringset_create(void)
+{
+    stringset_t *self = g_malloc0(sizeof *self);
```

Comment:
Nope. But in case of small allocations it is a sensible assumption. AFAIK use of overcommit -> the first sign of running out of memory will be segfault on memory access rather than getting an early warning as malloc / friends returning NULL -> preparing for NULL replies complicates things without really improving anything -> use of alloc or die functions is IMO justifiable.

Reply:
Yes. But it can die in a middle of some transaction. So we must be crash-proof then.